### DISCUSSION TITLE
Question on "cmake.sourceDirectory" substitution

### DISCUSSION THREAD

**Discussion by PennYU** [microsoft/vscode-cmake-tools#2637] [Q&A] [answered]
When I configure `settings.json` as:
![image](https://user-images.githubusercontent.com/15032921/177452781-5b5ee98f-6af2-49c9-b510-764f24a53d0b.png)

The variant for "cmake.buildDirectory" is substituted, but the variant for "cmake.sourceDirectory" isn't substituted:
![image](https://user-images.githubusercontent.com/15032921/177454450-83a0be5e-f22a-4d47-8f60-2ae192636777.png)

May i ask the purpose of the different substitution between "cmake.buildDirectory" and "cmake.sourceDirectory" ?
Thank you in advance.

**Comment by tigerlei010** [comment]
sometime user hope cmake tools can use different source directory according the cmake-variant choise. 
by reviewing source code , I found that expandString function use different argument to deal with sourceDirectory and buildDirectory, so I have the same question, why do this deliberate design ?

**Comment by bobbrow** [comment, marked as answer]
The design reflects our expectation that the source directory is relatively stable.  We haven't seen a project that changes the source directory based on the build type before.  Do you have separate projects for `src/Debug` and `src/Release`?  Do they eventually reference the same build scripts for the rest of the project?  We expected that most people would conditionally include a cmake rule file or use a toolchain to set project state based on build type instead.

We can consider a feature request to support this if you would like to open an issue for it.  There is some messiness with the build directory when source directory is unstable since we don't guarantee that the build directory is variable in the same way as the source directory, but perhaps we could enforce a condition that if variant substitution happens in the source directory, it must also happen for the build directory to avoid problems.

**Reply by PennYU** [reply]
Thank you so much. I actually tried to separate a quite large project to small projects, with a "CMakeLists.txt" for each small project. When I work on a small project, I could build it by select a variant for it. 

Fortunately, I found a better way for me today. With "Add Folder to Workspace...", I can add small projects to a workspace, and it works fine so far.

**Reply by bobbrow** [reply]
That sounds great.  At some point we will also support setting multiple paths for `cmake.sourceDirectory` as well (multiple projects per folder).  #1374

**Reply by PennYU** [reply]
Thank you so much. I would love to see this feature.
